File: ptcgp/engine/attack.py
"""Attack execution logic for the battle engine."""
from __future__ import annotations

from typing import Optional

from ptcgp.cards.database import get_card
from ptcgp.cards.types import CostSymbol
from ptcgp.engine.actions import SlotRef
from ptcgp.engine.constants import WEAKNESS_BONUS
from ptcgp.engine.state import GameState, PokemonSlot, StatusEffect


def can_pay_cost(slot: PokemonSlot, cost: tuple[CostSymbol, ...]) -> bool:
    """Return True if the slot has enough energy to pay the attack cost."""
    remaining = dict(slot.attached_energy)  # Element -> count
    colorless_needed = 0
    for symbol in cost:
        if symbol == CostSymbol.COLORLESS:
            colorless_needed += 1
        else:
            element = symbol.to_element()
            if remaining.get(element, 0) > 0:
                remaining[element] -= 1
            else:
                return False  # can't pay typed requirement
    total_remaining = sum(remaining.values())
    return total_remaining >= colorless_needed


def _retaliate_damage(defender_slot: PokemonSlot, defender_card) -> int:
    """Return the retaliate damage the defender deals back to the attacker.

    Reads both (a) the defender's passive ability text (e.g. Druddigon's
    "If this Pokémon is in the Active Spot and is damaged by an attack...")
    and (b) the attached tool card's effect text (e.g. Rocky Helmet). The
    retaliate only fires when the defender is currently in the Active Spot.
    """
    import re as _re
    pattern = _re.compile(
        r"is damaged by an attack.*do (\d+) damage to the attacking", _re.IGNORECASE
    )
    total = 0
    # Ability retaliate — only fires from the Active Spot per the card text.
    if defender_card.ability and defender_card.ability.effect_text:
        m = pattern.search(defender_card.ability.effect_text)
        if m:
            total += int(m.group(1))
    # Tool retaliate (Rocky Helmet).
    if defender_slot.tool_card_id:
        try:
            tool_card = get_card(defender_slot.tool_card_id)
        except KeyError:
            tool_card = None
        if tool_card and tool_card.trainer_effect_text:
            m = pattern.search(tool_card.trainer_effect_text)
            if m:
                total += int(m.group(1))
    return total


def execute_attack(
    state: GameState,
    attack_index: int,
    sub_target: Optional[SlotRef] = None,
) -> GameState:
    """Execute the current player's Active Pokemon's attack at ``attack_index``.

    Pipeline:
    1. Validate and pay cost
    2. Confusion check (may short-circuit)
    3. Compute final damage (base + weakness + parsed damage modifiers)
    4. Apply damage to defender (respecting "prevent damage" flags)
    5. Resolve side-effect handlers via the effect registry
    """
    state = state.copy()

    attacker_slot = state.players[state.current_player].active
    if attacker_slot is None:
        raise ValueError("Current player has no active Pokemon")

    attacker_card = get_card(attacker_slot.card_id)
    if attack_index < 0 or attack_index >= len(attacker_card.attacks):
        raise ValueError(f"Invalid attack_index {attack_index} for {attacker_card.name}")

    attack = attacker_card.attacks[attack_index]

    if not can_pay_cost(attacker_slot, attack.cost):
        raise ValueError(
            f"{attacker_card.name} cannot pay cost {attack.cost}: "
            f"has {attacker_slot.attached_energy}"
        )

    # Confusion check — tails means the attack fails entirely.
    if StatusEffect.CONFUSED in attacker_slot.status_effects:
        if state.rng.random() >= 0.5:
            return state

    defender_player = state.players[state.opponent_index]
    if defender_player.active is None:
        raise ValueError("Opponent has no active Pokemon")

    defender_slot = defender_player.active
    defender_card = get_card(defender_slot.card_id)

    # -- Phase 1: compute damage --------------------------------------- #
    base_damage = attack.damage
    if attacker_card.element is not None and defender_card.weakness == attacker_card.element:
        base_damage += WEAKNESS_BONUS

    final_damage = base_damage
    mod_skip = False
    modifier_result: dict = {}
    if attack.effect_text or attack.handler:
        from ptcgp.effects.damage_modifiers import compute_damage_modifier
        final_damage, mod_skip, modifier_result = compute_damage_modifier(
            state=state,
            base_damage=base_damage,
            raw_damage=attack.damage,
            attack_effect_text=attack.effect_text,
            attacker_slot=attacker_slot,
            attacker_card=attacker_card,
            defender_slot=defender_slot,
            defender_card=defender_card,
            handler_str=attack.handler,
            cached_effects=attack.cached_effects,
        )

    # -- Phase 2: apply damage ----------------------------------------- #
    # "Prevent all damage" flags set by previous-turn effects (e.g. coin-flip
    # damage block) take precedence and zero the hit.
    if defender_slot.prevent_damage_next_turn:
        final_damage = 0
        mod_skip = True

    damage_dealt = 0
    if not mod_skip and final_damage > 0:
        damage_dealt = min(final_damage, defender_slot.current_hp)
        defender_slot.current_hp = max(0, defender_slot.current_hp - final_damage)

    # -- Phase 2b: retaliate (Rocky Helmet / Druddigon-style passives) -- #
    if damage_dealt > 0:
        retaliate = _retaliate_damage(defender_slot, defender_card)
        if retaliate > 0:
            attacker_slot.current_hp = max(0, attacker_slot.current_hp - retaliate)

    # -- Phase 3: side-effect handlers --------------------------------- #
    if attack.effect_text or attack.handler:
        from ptcgp.effects.apply import apply_effects
        attacker_ref = SlotRef.active(state.current_player)
        target_ref = sub_target if sub_target is not None else SlotRef.active(state.opponent_index)
        extra = {"damage_dealt": damage_dealt, **modifier_result}
        state = apply_effects(
            state,
            attack.effect_text,
            acting_player=state.current_player,
            source_ref=attacker_ref,
            target_ref=target_ref,
            extra=extra,
            handler_str=attack.handler,
            cached_effects=attack.cached_effects,
        )

    return state

File: ptcgp/src/engine/attack.rs
//! Attack execution logic.
//!
//! Ported from `ptcgp/engine/attack.py`.

use std::collections::HashMap;
use std::sync::OnceLock;
use rand::Rng;
use regex::Regex;

use crate::card::{Card, CardDb};
use crate::state::{GameState, PokemonSlot};
use crate::actions::SlotRef;
use crate::types::{CostSymbol, StatusEffect};
use crate::constants::WEAKNESS_BONUS;
use crate::effects::{EffectContext, EffectKind};
use crate::effects::dispatch::{apply_effects, compute_damage_modifier};

#[inline]
/// Returns true if the slot has enough energy to pay the given cost.
///
/// Typed requirements are matched first; any leftover energy satisfies
/// Colorless requirements.
pub fn can_pay_cost(slot: &PokemonSlot, cost: &[CostSymbol]) -> bool {
    // Track remaining energy per element (copy of the slot's energy array).
    let mut remaining: [i8; 8] = [0; 8];
    for i in 0..8 {
        remaining[i] = slot.energy[i] as i8;
    }

    let mut colorless_needed: u8 = 0;

    for symbol in cost {
        match symbol {
            CostSymbol::Colorless => {
                colorless_needed += 1;
            }
            _ => {
                let el = symbol.to_element().expect("Non-colorless CostSymbol has element");
                if remaining[el as usize] > 0 {
                    remaining[el as usize] -= 1;
                } else {
                    return false; // can't pay typed requirement
                }
            }
        }
    }

    // Colorless can be paid by any remaining energy.
    let total_remaining: i8 = remaining.iter().sum();
    total_remaining >= colorless_needed as i8
}

/// Compute retaliate damage from Rocky Helmet / Druddigon-style passives.
///
/// Scans both the defender's ability text and attached tool text for the
/// pattern "is damaged by an attack...do X damage to the attacking".
fn retaliate_damage(defender_slot: &PokemonSlot, defender_card: &Card, db: &CardDb) -> i16 {
    // Compiled once at first call via OnceLock — avoids re-compiling on every attack.
    static RE: OnceLock<Regex> = OnceLock::new();
    let pattern = RE.get_or_init(|| {
        Regex::new(r"(?i)is damaged by an attack.*?do (\d+) damage to the attacking")
            .expect("Invalid retaliate regex")
    });

    let mut total: i16 = 0;

    // Ability retaliate (e.g. Druddigon).
    if let Some(ref ability) = defender_card.ability {
        if let Some(m) = pattern.captures(&ability.effect_text) {
            if let Some(n) = m.get(1).and_then(|s| s.as_str().parse::<i16>().ok()) {
                total += n;
            }
        }
    }

    // Tool retaliate (e.g. Rocky Helmet).
    if let Some(tool_idx) = defender_slot.tool_idx {
        let tool_card = db.get_by_idx(tool_idx);
        if let Some(m) = pattern.captures(&tool_card.trainer_effect_text) {
            if let Some(n) = m.get(1).and_then(|s| s.as_str().parse::<i16>().ok()) {
                total += n;
            }
        }
    }

    total
}

/// Execute attack at `attack_index` for the current player's active Pokemon.
///
/// `sub_target`: for targeted attacks; `None` = target opponent's active.
pub fn execute_attack(
    state: &mut GameState,
    db: &CardDb,
    attack_index: usize,
    sub_target: Option<SlotRef>,
    extra_target: Option<SlotRef>,
) {
    // 1. Validate attacker.
    let attacker_card_idx = state
        .players[state.current_player]
        .active
        .as_ref()
        .expect("Current player has no active Pokemon")
        .card_idx;

    let attacker_card = db.get_by_idx(attacker_card_idx);

    assert!(
        attack_index < attacker_card.attacks.len(),
        "Invalid attack_index {} for {}",
        attack_index,
        attacker_card.name
    );

    // 2. Check can_pay_cost.
    {
        let attacker_slot = state.players[state.current_player].active.as_ref().unwrap();
        let cost = &attacker_card.attacks[attack_index].cost;
        assert!(
            can_pay_cost(attacker_slot, cost),
            "{} cannot pay cost for attack {}",
            attacker_card.name,
            attacker_card.attacks[attack_index].name
        );
    }

    // 3. Confusion check: tails (>= 0.5) means the attack fails.
    let is_confused = state.players[state.current_player]
        .active
        .as_ref()
        .unwrap()
        .has_status(StatusEffect::Confused);
    if is_confused {
        let heads = state.rng.gen::<f64>() < 0.5;
        state.coin_flip_log.push(if heads {
            "🪙 Confusion flip: Heads! Attack succeeds".to_string()
        } else {
            // PTCGP confusion has NO self-damage on tails — attack just fails.
            "🪙 Confusion flip: Tails! Attack fails.".to_string()
        });
        if !heads {
            // The attack failed but still ends the turn.  The runner's dispatch
            // for ActionKind::Attack calls advance_turn after this returns,
            // so we do not call advance_turn directly here.
            return;
        }
    }

    // 4. Get defender (opponent's active).
    let opponent_idx = state.opponent_index();
    assert!(
        state.players[opponent_idx].active.is_some(),
        "Opponent has no active Pokemon"
    );

    // Clone cards for immutable access while we mutate state below.
    let attack = attacker_card.attacks[attack_index].clone();
    let attacker_element = attacker_card.element;
    let _attacker_card_clone = attacker_card.clone();

    let defender_card_idx = state.players[opponent_idx].active.as_ref().unwrap().card_idx;
    let defender_card = db.get_by_idx(defender_card_idx).clone();

    // 5. Compute base_damage with weakness bonus.
    // Only apply the +20 weakness bonus when the attack has a non-zero base
    // damage.  Zero-damage attacks (e.g. Tail Whip) must not deal bonus damage
    // simply because of a type match-up.
    let mut base_damage = attack.damage;
    if attack.damage > 0 && attacker_element.is_some() && defender_card.weakness == attacker_element {
        base_damage += WEAKNESS_BONUS;
    }

    // 6. Compute damage modifier.
    let ctx = EffectContext {
        acting_player: state.current_player,
        source_ref: Some(SlotRef::active(state.current_player)),
        target_ref: sub_target,
        extra_target_ref: extra_target,
        extra: HashMap::new(),
    };
    // Use pre-parsed attack effects (parsed at card load time).
    let attack_effects = attack.effects.clone();
    let (mut final_damage, mod_skip, modifier_result) =
        compute_damage_modifier(state, db, base_damage, &attack_effects, &ctx);

    // 7. Respect "prevent damage" flag.
    let prevent = state.players[opponent_idx]
        .active
        .as_ref()
        .unwrap()
        .prevent_damage_next_turn;
    let mod_skip = if prevent {
        final_damage = 0;
        true
    } else {
        mod_skip
    };

    // 7b. PassivePreventExDamage (Oricorio Safeguard): if the defender's active
    // ability has this passive and the attacker is an ex Pokemon, negate damage.
    let attacker_is_ex = db.get_by_idx(attacker_card_idx).is_ex;
    if attacker_is_ex && !mod_skip {
        let defender_has_safeguard = state.players[opponent_idx]
            .active
            .as_ref()
            .map(|s| {
                let card = db.get_by_idx(s.card_idx);
                card.ability.as_ref().map(|ab| {
                    ab.effects.iter().any(|e| matches!(e, EffectKind::PassivePreventExDamage))
                }).unwrap_or(false)
            })
            .unwrap_or(false);
        if defender_has_safeguard {
            final_damage = 0;
        }
    }

    // 8. Apply damage — first subtract the defender's incoming_damage_reduction
    // (set by effects like PassiveIncomingDamageReduction), clamped at 0.
    if !mod_skip && final_damage > 0 {
        let reduction = state.players[opponent_idx]
            .active
            .as_ref()
            .unwrap()
            .incoming_damage_reduction as i16;
        if reduction > 0 {
            final_damage = (final_damage - reduction).max(0);
        }
    }

    let damage_dealt: i16;
    let opponent_ko: bool;
    if !mod_skip && final_damage > 0 {
        let defender_hp = state.players[opponent_idx].active.as_ref().unwrap().current_hp;
        damage_dealt = final_damage.min(defender_hp);
        let new_hp = (defender_hp - final_damage).max(0);
        state.players[opponent_idx].active.as_mut().unwrap().current_hp = new_hp;
        opponent_ko = new_hp == 0;
    } else {
        damage_dealt = 0;
        opponent_ko = false;
    }

    // 9. Retaliate (Rocky Helmet / Druddigon).
    if damage_dealt > 0 {
        let defender_slot_ref = state.players[opponent_idx].active.as_ref().unwrap();
        let retaliate = retaliate_damage(defender_slot_ref, &defender_card, db);
        if retaliate > 0 {
            let attacker = state.players[state.current_player].active.as_mut().unwrap();
            attacker.current_hp = (attacker.current_hp - retaliate).max(0);
        }

        // Poison Barb: if defender's tool has PassiveRetaliatePoison, poison the attacker.
        let has_poison_barb = state.players[opponent_idx].active.as_ref()
            .and_then(|slot| slot.tool_idx)
            .and_then(|tidx| db.try_get_by_idx(tidx))
            .map(|tool| {
                tool.trainer_effects.iter().any(|e| matches!(e, EffectKind::PassiveRetaliatePoison))
            })
            .unwrap_or(false);
        if has_poison_barb {
            if let Some(attacker) = state.players[state.current_player].active.as_mut() {
                attacker.add_status(StatusEffect::Poisoned);
            }
        }
    }

    // 10. Apply side-effect handlers.
    let attacker_ref = SlotRef::active(state.current_player);
    let ctx_with_damage = EffectContext {
        acting_player: state.current_player,
        source_ref: Some(attacker_ref),
        target_ref: sub_target,
        extra_target_ref: extra_target,
        extra: {
            let mut m = modifier_result
                .iter()
                .map(|(k, &v)| (k.clone(), v))
                .collect::<HashMap<String, i32>>();
            m.insert("damage_dealt".to_string(), damage_dealt as i32);
            m.insert("opponent_ko".to_string(), opponent_ko as i32);
            m
        },
    };

    // Apply post-damage side effects (status conditions, splash, heal, etc.).
    apply_effects(state, db, &attack_effects, &ctx_with_damage);

    // Resolve any KOs caused by the attack (or by retaliate damage) so the
    // state is consistent before the runner advances the turn.
    crate::engine::ko::check_and_handle_kos(state, db);

    // PTCGP rule: an attack always ends the attacker's turn.  If the KO
    // pushed us into AwaitingBenchPromotion, the runner can't `advance_turn`
    // immediately — the defender must promote first.  Set a flag so the
    // runner / MCTS knows to advance after promotion completes.
    if state.phase == crate::types::GamePhase::AwaitingBenchPromotion {
        state.attack_pending_advance = true;
    }
}

// ------------------------------------------------------------------ //
// Tests
// ------------------------------------------------------------------ //

#[cfg(test)]
mod tests {
    use super::*;
    use crate::state::PokemonSlot;
    use crate::types::Element;

    fn make_slot(hp: i16) -> PokemonSlot {
        PokemonSlot::new(0, hp)
    }

    #[test]
    fn can_pay_cost_no_cost() {
        let slot = make_slot(100);
        assert!(can_pay_cost(&slot, &[]));
    }

    #[test]
    fn can_pay_cost_typed_exact() {
        let mut slot = make_slot(100);
        slot.add_energy(Element::Fire, 2);
        let cost = vec![CostSymbol::Fire, CostSymbol::Fire];
        assert!(can_pay_cost(&slot, &cost));
    }

    #[test]
    fn can_pay_cost_typed_insufficient() {
        let mut slot = make_slot(100);
        slot.add_energy(Element::Fire, 1);
        let cost = vec![CostSymbol::Fire, CostSymbol::Fire];
        assert!(!can_pay_cost(&slot, &cost));
    }

    #[test]
    fn can_pay_cost_colorless_satisfied_by_any_energy() {
        let mut slot = make_slot(100);
        slot.add_energy(Element::Water, 1);
        let cost = vec![CostSymbol::Colorless];
        assert!(can_pay_cost(&slot, &cost));
    }

    #[test]
    fn can_pay_cost_colorless_insufficient() {
        let slot = make_slot(100);
        let cost = vec![CostSymbol::Colorless];
        assert!(!can_pay_cost(&slot, &cost));
    }

    #[test]
    fn can_pay_cost_mixed_typed_and_colorless() {
        let mut slot = make_slot(100);
        slot.add_energy(Element::Grass, 1);
        slot.add_energy(Element::Water, 1);
        // Cost: 1 Grass + 1 Colorless
        let cost = vec![CostSymbol::Grass, CostSymbol::Colorless];
        assert!(can_pay_cost(&slot, &cost));
    }

    #[test]
    fn can_pay_cost_typed_cannot_satisfy_colorless_from_wrong_type() {
        let mut slot = make_slot(100);
        // Has exactly 1 Fire — enough for 1 Fire, but not also 1 Water requirement.
        slot.add_energy(Element::Fire, 1);
        let cost = vec![CostSymbol::Fire, CostSymbol::Water];
        assert!(!can_pay_cost(&slot, &cost));
    }
}
